In [373]:
# Get imports
import pandas as pd
import numpy as np 



In [374]:
# Let's make a function that takes in a database name as a string
# this function also performs our imports from env
def get_db_url(db_name):
    from env import user, host, password
    return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'

In [375]:
sql = """
    SELECT * from employees
"""

# 5
Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?

    - NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:amysql.pymysql

b. Intentionally make an error in your SQL query. What does the error message look like?

    - ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'fro employees limit 3000' at line 1")
[SQL: 
    SELECT * fro employees limit 3000
]
(Background on this error at: http://sqlalche.me/e/14/f405)

# 6 
Read the employees and titles tables into two separate DataFrames.

In [376]:
query_6_1 = '''
SELECT * 
FROM employees
LIMIT 3000
'''

query_6_2 = '''
SELECT * 
FROM titles
LIMIT 3000
'''

In [377]:
url = get_db_url("employees")
df_6_employees = pd.read_sql(query_6_1, url)
df_6_employees

emp_no  birth_date first_name last_name gender   hire_date
0      10001  1953-09-02     Georgi   Facello      M  1986-06-26
1      10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2      10003  1959-12-03      Parto   Bamford      M  1986-08-28
3      10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4      10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...      ...         ...        ...       ...    ...         ...
2995   12996  1964-12-07     Khalil   Madeira      F  1992-07-14
2996   12997  1956-09-22       Zine     Swist      F  1991-03-22
2997   12998  1956-09-02     Nadjib   Rindone      M  1991-01-23
2998   12999  1963-10-13      Taizo    Koyama      F  1991-11-06
2999   13000  1955-01-31    Adhemar    Chiodo      F  1994-05-11

[3000 rows x 6 columns]

In [378]:
df_6_titles = pd.read_sql(query_6_2, url)
df_6_titles

emp_no            title   from_date     to_date
0      10001  Senior Engineer  1986-06-26  9999-01-01
1      10002            Staff  1996-08-03  9999-01-01
2      10003  Senior Engineer  1995-12-03  9999-01-01
3      10004         Engineer  1986-12-01  1995-12-01
4      10004  Senior Engineer  1995-12-01  9999-01-01
...      ...              ...         ...         ...
2995   12019     Senior Staff  1991-02-17  9999-01-01
2996   12019            Staff  1986-02-17  1991-02-17
2997   12020     Senior Staff  1993-08-09  9999-01-01
2998   12020            Staff  1988-08-09  1993-08-09
2999   12021         Engineer  1991-09-30  2000-09-29

[3000 rows x 4 columns]

# 7 
How many rows and columns do you have in each DataFrame? Is that what you expected?

    - Employees 300,0024 rows x 6 columns
    - Titles 443,308 rows x 4 columns 
    Yes, I was expecting the titles table to have more records because one person can habe multiple jobs 
    



# 8 
Display the summary statistics for each DataFrame

In [379]:
df_6_employees.describe()

emp_no
count   3000.000000
mean   11500.500000
std      866.169729
min    10001.000000
25%    10750.750000
50%    11500.500000
75%    12250.250000
max    13000.000000

In [380]:
df_6_titles.describe()

emp_no
count   3000.000000
mean   11013.785000
std      583.675637
min    10001.000000
25%    10511.000000
50%    11021.500000
75%    11518.000000
max    12021.000000

# 9 
How many unique titles are in the titles DataFrame?

In [381]:
df_6_titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   emp_no     3000 non-null   int64 
 1   title      3000 non-null   object
 2   from_date  3000 non-null   object
 3   to_date    3000 non-null   object
dtypes: int64(1), object(3)
memory usage: 93.9+ KB


In [382]:
df_6_titles.title.unique()

array(['Senior Engineer', 'Staff', 'Engineer', 'Senior Staff',
       'Assistant Engineer', 'Technique Leader'], dtype=object)

# 10
What is the oldest date in the to_date column?

In [383]:
df_6_titles.to_date.min()

datetime.date(1986, 3, 7)

# 11
What is the most recent date in the to_date column?


In [384]:
df_6_titles.to_date.max()

datetime.date(9999, 1, 1)

# 

#  <center>Exercises II</center></h1>

# 2.1
Copy the users and roles DataFrames from the examples above.

In [385]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

In [386]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

# 2.2 
What is the result of using a right join on the DataFrames?

In [387]:
right_join = pd.merge(users, roles, left_on='role_id', right_on='id', how='right')
right_join

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0     1      admin
1   2.0    joe      2.0     2     author
2   3.0  sally      3.0     3   reviewer
3   4.0   adam      3.0     3   reviewer
4   NaN    NaN      NaN     4  commenter

# 2.3
What is the result of using an outer join on the DataFrames?

In [388]:
left_join = pd.merge(users, roles, left_on='role_id', right_on='id', how='outer')
left_join

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0   1.0      admin
1   2.0    joe      2.0   2.0     author
2   3.0  sally      3.0   3.0   reviewer
3   4.0   adam      3.0   3.0   reviewer
4   5.0   jane      NaN   NaN        NaN
5   6.0   mike      NaN   NaN        NaN
6   NaN    NaN      NaN   4.0  commenter

# 2.4
What happens if you drop the foreign keys from the DataFrames and try to merge them?

- without the foreing key one can not connect two databases correctly 

# 2. 5
Load the mpg dataset from PyDataset.

In [389]:
from pydataset import data
mpg = data('mpg')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

# 2.6
Output and read the documentation for the mpg dataset.

In [390]:
data('mpg', show_doc=True) 

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




# 2.7
How many rows and columns are in the dataset?

There are 11 columns with 234 records

In [391]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


# 2.8
Check out your column names and perform any cleanup you may want on them  

In [392]:
mpg.columns

Index(['manufacturer', 'model', 'displ', 'year', 'cyl', 'trans', 'drv', 'cty',
       'hwy', 'fl', 'class'],
      dtype='object')

In [393]:
mpg.head(3)

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact

In [394]:
mpg = mpg.rename(columns={"cty": "city"})

# 2.9
Display the summary statistics for the dataset. 

In [395]:
mpg.describe()

displ         year         cyl        city         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

# 2.10
How many different manufacturers are there? 

In [396]:
len((mpg['manufacturer'].unique()))

15

# 2.11
How many different models are there?

In [397]:
models = pd.DataFrame(mpg['model'].value_counts())
models.shape[0]

38

In [398]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  city  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f    18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f    21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f    20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f    21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f    16   26  p  compact

# 2.12
Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [399]:
mpg['mileage_difference'] = mpg.hwy - mpg.city
mpg

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

# 2.13
Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [400]:
mpg['average_mileage'] = ((mpg.city + mpg.hwy) / 2)
mpg

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

# 2.14
Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission. 

In [401]:
# #mpg[mpg['trans'].astype('str') == mpg['']]
# trans = mpg['trans'][1][0] == 'a'
# # trans[1][0] == 'a'
# trans
mpg['is_automatic'] = mpg.trans.apply(lambda value : value[0] == 'a')
mpg


manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  mileage_difference  average_mileage  is_automatic  
1    compact                  11             23.5          True  
2    compact                   8             25.0         False  
3    compact                  11             25.5         False  
4    compact                   9             25.5          True  
5    compact                  10             21.0          True  
..       ...                 ...              ...           ...  
230  midsize                   9             23.5          True  
231  midsize                   8             25.0         False  
232  midsize                  10             21.0          True  
233  midsize                   8             22.0         False  
234  midsize                   9             21.5          True  

[234 rows x 14 columns]

#  2.15
Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average? 

In [402]:
# mpg[mpg.groupby('manufacturer').average_mileage.agg('mean') == mpg.groupby('manufacturer').average_mileage.agg('mean').max()]
best = mpg.groupby('manufacturer').average_mileage.agg('mean') == mpg.groupby('manufacturer').average_mileage.agg('mean').max()
best[best == True]

manufacturer
honda    True
Name: average_mileage, dtype: bool

# 2.16
Do automatic or manual cars have better miles per gallon? 

In [403]:
automatic = mpg[mpg['is_automatic'] == True]
automatic.head()

manufacturer       model  displ  year  cyl     trans drv  city  hwy fl  \
1         audi          a4    1.8  1999    4  auto(l5)   f    18   29  p   
4         audi          a4    2.0  2008    4  auto(av)   f    21   30  p   
5         audi          a4    2.8  1999    6  auto(l5)   f    16   26  p   
7         audi          a4    3.1  2008    6  auto(av)   f    18   27  p   
9         audi  a4 quattro    1.8  1999    4  auto(l5)   4    16   25  p   

     class  mileage_difference  average_mileage  is_automatic  
1  compact                  11             23.5          True  
4  compact                   9             25.5          True  
5  compact                  10             21.0          True  
7  compact                   9             22.5          True  
9  compact                   9             20.5          True

In [404]:
automatic.average_mileage.agg('mean')

19.130573248407643

In [405]:
manual = mpg[mpg['is_automatic'] == False]
manual.head()

manufacturer       model  displ  year  cyl       trans drv  city  hwy fl  \
2          audi          a4    1.8  1999    4  manual(m5)   f    21   29  p   
3          audi          a4    2.0  2008    4  manual(m6)   f    20   31  p   
6          audi          a4    2.8  1999    6  manual(m5)   f    18   26  p   
8          audi  a4 quattro    1.8  1999    4  manual(m5)   4    18   26  p   
10         audi  a4 quattro    2.0  2008    4  manual(m6)   4    20   28  p   

      class  mileage_difference  average_mileage  is_automatic  
2   compact                   8             25.0         False  
3   compact                  11             25.5         False  
6   compact                   8             22.0         False  
8   compact                   8             22.0         False  
10  compact                   8             24.0         False

In [406]:
manual.average_mileage.agg('mean')


22.227272727272727

In [407]:
print(f"Manual cars get an average mileage per galon consumption of {round(manual.average_mileage.agg('mean'), 2)}")
print(f"Automatic cars get {round(automatic.average_mileage.agg('mean'), 2)}")
print(f"Manual cars get {round(manual.average_mileage.agg('mean') - round(automatic.average_mileage.agg('mean')), 2)} miles per galon more")

Manual cars get an average mileage per galon consumption of 22.23
Automatic cars get 19.13
Manual cars get 3.23 miles per galon more


# 

 #  <center>Exercises III</center></h1>

# 3.1
Use your get_db_url function to help you explore the data from the chipotle database.

In [408]:
def get_db_url(db_name):
    from env import user, host, password
    return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'
    
url_c = get_db_url("chipotle")
sql_c = """
    SELECT * from orders

"""

url = get_db_url("employees")
df_c = pd.read_sql(sql_c, url_c)
df_c


id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

# 3.2
What is the total price for each order?

In [411]:
order_price = df_c.groupby('order_id').item_price.agg('sum')
order_price

order_id
1       11.56
2       16.98
3       12.67
4       21.00
5       13.70
        ...  
1830    23.00
1831    12.90
1832    13.20
1833    23.50
1834    28.75
Name: item_price, Length: 1834, dtype: float64

# 3.3
What are the most popular 3 items?

In [428]:
fav_items = df_c.groupby('item_name').quantity.agg('sum').sort_values(ascending=False)
fav_items.head(3)

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

# 3.4
Which item has produced the most revenue?

In [455]:
item_price = df_c.groupby('item_name').item_price.agg('sum')
item_price[item_price == item_price.max()]

item_name
Chicken Bowl    7342.73
Name: item_price, dtype: float64

# 3.5
Join the employees and titles DataFrames together.

In [458]:
employees = df_6_employees
titles = df_6_titles 

In [460]:
employees.head(3)

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28

In [461]:
titles.head(3)

emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01

In [469]:
employee_title = pd.merge(employees, titles, on='emp_no')
print(employee_title.info())
print(employee_title.head(3))
print(type(employee_title['hire_date'][1]))


<class 'pandas.core.frame.DataFrame'>
Int64Index: 3000 entries, 0 to 2999
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   emp_no      3000 non-null   int64 
 1   birth_date  3000 non-null   object
 2   first_name  3000 non-null   object
 3   last_name   3000 non-null   object
 4   gender      3000 non-null   object
 5   hire_date   3000 non-null   object
 6   title       3000 non-null   object
 7   from_date   3000 non-null   object
 8   to_date     3000 non-null   object
dtypes: int64(1), object(8)
memory usage: 234.4+ KB
None
   emp_no  birth_date first_name last_name gender   hire_date  \
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28   

             title   from_date     to_date  
0  Senior Engineer  1986-06-26  9999-01-01  
1            Staff  1996-08-03  9999-01-01  
2

# 3.6
For each title, find the hire date of the employee that was hired most recently with that title.

In [474]:
by_title = employee_title.groupby('title').hire_date.max()
by_title

title
Assistant Engineer    1999-02-17
Engineer              1999-10-28
Senior Engineer       1998-11-03
Senior Staff          1998-06-18
Staff                 1999-08-23
Technique Leader      1999-11-06
Name: hire_date, dtype: object

In [ ]:
best = mpg.groupby('manufacturer').average_mileage.agg('mean') == mpg.groupby('manufacturer').average_mileage.agg('mean').max()

# 3.7
Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)   

SyntaxError: unexpected character after line continuation character (<ipython-input-475-1b2b9feacc24>, line 1)